src/backend/gl/src/queue.rs

```diff
@@ -730,12 +730,58 @@ impl hal::queue::RawCommandQueue<Backend> for CommandQueue {
     {
         use glutin::GlContext;
 
-        for swapchain in swapchains {
-            swapchain.0
-                .borrow()
+        for (swapchain, index) in swapchains {
+            let swapchain = swapchain.borrow();
+
+            // NOTE: If the extents don't match, it is *probably only* suboptimal.
+            // HAL doesn't currently distinguish between suboptimal and out of date,
+            // so we just have to return Err. This is a note to whoever adds the
+            // ability to distinquish between the two.
+            //
+            // NOTE: I say probably because I'm not sure what happens if you
+            // the dimmensions of the destination fbo you give is bigger than
+            // the actual dimmensions of the destination fbo. It's probably ok,
+            // but you should probably check.
+            let extent = window::glutin::physical_to_extent(
+                &swapchain
+                .surface
+                .surface_context
                 .window
+                .get_inner_size()
+                .unwrap()
+                .to_physical(swapchain.surface.surface_context.window.get_hidpi_factor())
+            );
+
+            if extent != swapchain.extent {
+                return Err(());
+            }
+
+            unsafe {
+                let gl = &swapchain.surface.device.share.context;
+                let fbo = swapchain.images[index as usize].fbo;
+                gl.BindFramebuffer(gl::READ_FRAMEBUFFER, fbo);
```

Comment:
Won't work with OpenGL 2.1 without the `EXT_framebuffer_object` and `EXT_framebuffer_bind` extensions. Even then we would have to append `EXT` to all the relevant functions and `_EXT` to all the relevant constants.